lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -21,15 +25,148 @@ using namespace mlir;
 
 namespace circt {
 
+struct ModuleOpConversion : public OpRewritePattern<mlir::ModuleOp> {
+  ModuleOpConversion(MLIRContext *context, StringRef topLevelFunction,
+                     calyx::ProgramOp *programOpOutput)
+      : OpRewritePattern<mlir::ModuleOp>(context),
+        programOpOutput(programOpOutput), topLevelFunction(topLevelFunction) {
+    assert(programOpOutput->getOperation() == nullptr &&
+           "this function will set programOpOutput post module conversion");
+  }
+
+  LogicalResult matchAndRewrite(mlir::ModuleOp moduleOp,
+                                PatternRewriter &rewriter) const override {
+    if (!moduleOp.getOps<calyx::ProgramOp>().empty())
+      return failure();
+
+    rewriter.updateRootInPlace(moduleOp, [&] {
+      // Create ProgramOp
+      rewriter.setInsertionPointAfter(moduleOp);
+      auto programOp = rewriter.create<calyx::ProgramOp>(
+          moduleOp.getLoc(), StringAttr::get(getContext(), topLevelFunction));
+
+      // Inline the module body region
+      rewriter.inlineRegionBefore(moduleOp.getBodyRegion(),
+                                  programOp.getBodyRegion(),
+                                  programOp.getBodyRegion().end());
+
+      // Inlining the body region also removes ^bb0 from the module body
+      // region, so recreate that, before finally inserting the programOp
+      auto moduleBlock = rewriter.createBlock(&moduleOp.getBodyRegion());
+      rewriter.setInsertionPointToStart(moduleBlock);
+      rewriter.insert(programOp);
+      *programOpOutput = programOp;
+    });
+    return success();
+  }
+
+private:
+  calyx::ProgramOp *programOpOutput = nullptr;
+  StringRef topLevelFunction;
+};
+
 //===----------------------------------------------------------------------===//
 // Pass driver
 //===----------------------------------------------------------------------===//
 class SCFToCalyxPass : public SCFToCalyxBase<SCFToCalyxPass> {
 public:
   SCFToCalyxPass() : SCFToCalyxBase<SCFToCalyxPass>() {}
-  void runOnOperation() override{};
+  void runOnOperation() override;
+
+  LogicalResult setTopLevelFunction(mlir::ModuleOp moduleOp,
+                                    std::string &topLevelFunction) {
+    if (!topLevelFunctionOpt.empty()) {
+      if (SymbolTable::lookupSymbolIn(moduleOp, topLevelFunctionOpt) ==
+          nullptr) {
+        moduleOp.emitError() << "Top level function '" << topLevelFunctionOpt
+                             << "' not found in module.";
+        return failure();
+      }
+      topLevelFunction = topLevelFunctionOpt;
+    } else {
+      /// No top level function set; infer top level if the module only contains
+      /// a single function, else, throw error.
+      auto funcOps = moduleOp.getOps<mlir::FuncOp>();
+      if (std::distance(funcOps.begin(), funcOps.end()) == 1)
+        topLevelFunction = (*funcOps.begin()).sym_name().str();
```

Comment:
Nit: May be wrong, but possibly could be simplified to `.front().sym_name().str()`.

Reply:
`::front` is unfortunately not available for `iterator_range` :(.